fms/modules/tp.py

```diff
@@ -31,54 +25,229 @@ def setup_tp(self, rank: int, world_size: int) -> None:
+                [
+                    tensor_value[
+                        :,
+                        tp_slice,
+                    ]
+                    for tp_slice in self.__get_tp_slices(
+                        tensor_value, output_size_per_partition, max_partition_sizes
+                    )
+                ],
+                dim=1,
+            )
+            param.copy_(tensor, non_blocking=True)
+        else:
+            if self.rank == 0:
+                param.copy_(tensor_value, non_blocking=True)
+            else:
+                param.zero_()
+
+    def copy_colwise(
+        self,
+        param: Union[torch.nn.Parameter, torch.Tensor],
+        tensor_value: torch.Tensor,
+        is_bias: bool,
+        max_partition_sizes: List[int],
+    ):
+        """
+        This function copies the correct shard of the weights for a colwise-TP'd module
+        according to the rank of the process and the world_size.
+
+        Args
+        ====
+        param: torch.nn.Parameter
+            Parameter that has had TP applied
+        tensor_value: torch.Tensor
+            tensor that needs sharding
+        is_bias: bool
+            if True is bias, else is weight
+        max_partition_sizes: List[int]
+            for each number in the list, if world_size is smaller than or equal to that number, the tensor will get
+            partitioned in worldsize parts, else if world size is larger than the number then you will get world size parts
+            replicated in worldsize / number
+
+            world_size = 4, max_partition_sizes = [8], tensor = [0 1 2 3 4 5 6 7]
+            [0 1] [2 3] [4 5] [6 7]
 
-    def import_weights(self, module: nn.Module):
-        for weight in self.colwise_param_names():
-            apply_colwise_tp(
-                getattr(self, weight),
-                getattr(module, weight),
-                self.world_size,
-                self.rank,
+            world_size = 8, max_partition_sizes = [4], tensor = [0 1 2 3 4 5 6 7]
+            [0 1] [0 1] [2 3] [2 3] [4 5] [4 5] [6 7] [6 7]
+
+            If there are multiple numbers in the max_partition_sizes list, then the param gets filled with non-contiguous
+            slices of the tensor_value. This is useful for fused weight cases (qkv, mlp, moe, etc.)
+
+            world_size = 4, max_partition_sizes = [4, 4], tensor = [0 1 2 3 4 5 6 7]
+            [0 4] [1 5] [2 6] [3 7]
+
+            world_size = 4, max_partition_sizes = [4, 1], tensor = [0 1 2 3 4 5 6 7 8 9]
+            [0 1 8 9] [2 3 8 9] [4 5 8 9] [6 7 8 9]
+        """
+        # Divide the weight matrix along the first dimension.
+        output_size_per_partition = param.shape[0] // (
+            sum(max_partition_sizes) // min(max_partition_sizes)
+        )
+        if not is_bias:
+            tensor = torch.cat(
+                [
+                    tensor_value[
+                        tp_slice,
+                        :,
+                    ]
```

Comment:
also here I think we can clean up the formatting with a couple of variables to hold the tensors-to-concat